### PR TITLE
Do not exclude exit code -1 from restart

### PR DESCRIPTION
**Pull request by mhoeger**

### CODE REVIEW

**Comment by mhoeger** [comment]
@fabiocav - after talking to @amamounelsayed, we realized that the -1 was likely hard-coded in there because the java worker explicitly sets this exit code when it runs into an unhandled exception. However, we do want to restart the worker in these cases. It doesn't feel as risky because in node, we do the same thing (restart process on unhandled exception. Will test change in a private stamp too with placeholders.

Do you mind reviewing the code?

**Comment by amamounelsayed** [comment]
https://github.com/Azure/azure-functions-java-worker/blob/4fdc30170610393dca33a40725f1820340429376/src/main/java/com/microsoft/azure/functions/worker/Application.java#L123-L127

This is the line that calls the grpc send there is no handling exception around it just throws it to to catch all in Application class.
https://github.com/Azure/azure-functions-java-worker/blob/4fdc30170610393dca33a40725f1820340429376/src/main/java/com/microsoft/azure/functions/worker/JavaWorkerClient.java#L109

File: src/WebJobs.Script/Workers/Http/HttpWorkerProcess.cs [hunk 1]
```diff
@@ -54,11 +54,8 @@ internal override Process CreateWorkerProcess()
         internal override void HandleWorkerProcessExitError(WorkerProcessExitException langExc)
         {
             // The subscriber of WorkerErrorEvent is expected to Dispose() the errored channel
-            if (langExc != null && langExc.ExitCode != -1)
-            {
-                _workerProcessLogger.LogDebug(langExc, $"Language Worker Process exited.", _workerProcessArguments.ExecutablePath);
-                _eventManager.Publish(new HttpWorkerErrorEvent(_workerId, langExc));
-            }
+            _workerProcessLogger.LogDebug(langExc, $"Language Worker Process exited. Pid={langExc.Pid}.", _workerProcessArguments.ExecutablePath);
```
Comment (amamounelsayed) [inline src/WebJobs.Script/Workers/Http/HttpWorkerProcess.cs]: Are we always sure that we have langExc  value?

File: src/WebJobs.Script/Workers/Rpc/RpcWorkerProcess.cs [hunk 2]
```diff
@@ -50,11 +50,8 @@ internal override Process CreateWorkerProcess()
         internal override void HandleWorkerProcessExitError(WorkerProcessExitException langExc)
         {
             // The subscriber of WorkerErrorEvent is expected to Dispose() the errored channel
-            if (langExc != null && langExc.ExitCode != -1)
-            {
-                _workerProcessLogger.LogDebug(langExc, $"Language Worker Process exited.", _workerProcessArguments.ExecutablePath);
-                _eventManager.Publish(new WorkerErrorEvent(_runtime, _workerId, langExc));
-            }
+            _workerProcessLogger.LogDebug(langExc, $"Language Worker Process exited. Pid={langExc.Pid}.", _workerProcessArguments.ExecutablePath);
```
Comment (amamounelsayed) [inline src/WebJobs.Script/Workers/Rpc/RpcWorkerProcess.cs]: Are we always sure that we have langExc value?

File: src/WebJobs.Script/Workers/Http/HttpWorkerProcess.cs (hunk 1 above)
Comment (fabiocav) [reply src/WebJobs.Script/Workers/Http/HttpWorkerProcess.cs]: The fact that we had a check before leads me to believe we wanted to handle the possibility that this was `null`. If that isn't the case, we should have a proper argument validation here.

Comment (mhoeger) [reply src/WebJobs.Script/Workers/Http/HttpWorkerProcess.cs]: This method is only called here, where an exception is explicitly created: https://github.com/Azure/azure-functions-host/blob/dev/src/WebJobs.Script/Workers/ProcessManagement/WorkerProcess.cs#L139

Adding argument validation!

Comment (fabiocav) [reply src/WebJobs.Script/Workers/Http/HttpWorkerProcess.cs]: Yeah, if the expectation is that this wouldn't be called without a `null` argument, let's have this method validate that

File: src/WebJobs.Script/Workers/Rpc/RpcWorkerProcess.cs (hunk 2 above)
Comment (fabiocav) [reply src/WebJobs.Script/Workers/Rpc/RpcWorkerProcess.cs]: Add validation here as well.

File: src/WebJobs.Script/Workers/Http/HttpWorkerProcess.cs (hunk 1 above)
Comment (mhoeger) [reply src/WebJobs.Script/Workers/Http/HttpWorkerProcess.cs]: @fabiocav and @amamounelsayed - thoughts now? also if you have a better suggestion for the ArgumentNullException input very much appreciated!

Comment (fabiocav) [reply src/WebJobs.Script/Workers/Http/HttpWorkerProcess.cs]: The expectation is that you'd pass the argument name into that constructor.

Change to :
``` csharp
throw new ArgumentNullException(nameof(langExc));
```

Though, a better argument name would be nice :)

Comment (mhoeger) [reply src/WebJobs.Script/Workers/Http/HttpWorkerProcess.cs]: thanks!! do you mind reviewing again?

**Comment by mhoeger** [comment]
@fabiocav - one issue (that has always existed) @amamounelsayed found while testing is that on worker crash, we don't set the functions that were executing at that point to fail. The executions that were running when the crash happened will later cause function timeout exceptions.
https://github.com/Azure/azure-functions-host/issues/5385

This PR still gets us in a better place, but we wanted to give you a heads up!

**Comment by pragnagopa** [comment]
Apologies for pinging on a closed PR. Exit code -1 only happens if functions host shutsdown the child process either as part of jobhost restarts or due to a an unhandled exception

Restarting on exit code -1 would essentially be a no-op and causes more confusing logs.

**Comment by mhoeger** [comment]
@pragnagopa - agreed that worker shouldn't exit on error code -1! thanks! I do think though that ignoring exit code -1's is dangerous though since there's no guarantee that they only come from external killing due to jobhost restart. Do you think it would make sense to include some sort of check to see if we're disposing before starting a new process [here](https://github.com/Azure/azure-functions-host/blob/dev/src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs#L304) instead?

**Comment by pragnagopa** [comment]
There is already a check [here](https://github.com/Azure/azure-functions-host/blob/dev/src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs#L297) which I believe should help.

Regarding exit code '-1',  setting to a exit code < 0 is not by design. Here is documentation for Unix/Linux: https://shapeshed.com/unix-exit-codes/#how-to-set-an-exit-code. I did not find anything explicit for windows.

Here is a repro:

```
using System;
using System.Diagnostics;

namespace ConsoleApp1
{
    class Program
    {
        public static Process _process;
        static void Main(string[] args)
        {
            Console.WriteLine("Hello World!");
            var startInfo = new ProcessStartInfo("notepad.exe")
            {
                RedirectStandardOutput = true,
                RedirectStandardError = true,
                CreateNoWindow = true,
                UseShellExecute = false,
                ErrorDialog = false
            };

            _process = new Process { StartInfo = startInfo };
            try
            {
                _process.ErrorDataReceived += (sender, e) => OnErrorDataReceived(sender, e);
                _process.Exited += (sender, e) => OnProcessExited(sender, e);
                _process.EnableRaisingEvents = true;

                _process.Start();

                _process.BeginErrorReadLine();
                _process.BeginOutputReadLine();
            }
            catch (Exception ex)
            {
            }

            Console.WriteLine("Press anykey to dispose...");
            Console.ReadKey();
            Dispose();
        }

        private static void OnErrorDataReceived(object sender, DataReceivedEventArgs e)
        {
            // TODO: per language stdout/err parser?
        }

        private static void OnProcessExited(object sender, EventArgs e)
        {
            try
            {
                if (_process.ExitCode == -1)
                {
                    _process.WaitForExit();
                    _process.Close();
                }
            }
            catch (Exception exc)
            {
                // ignore process is already disposed
            }
        }

        public static void Dispose()
        {
            // best effort process disposal
            try
            {
                if (_process != null)
                {
                    if (!_process.HasExited)
                    {
                        _process.Kill();
                        _process.WaitForExit();
                    }
                    _process.Dispose();
                }
            }
            catch (Exception exc)
            {
                //ignore
            }
        }
    }
}

```
Exit code when dispose is explicitly called in `-1`. Unfortunately, there is not official documentation on this.

> ignoring exit code -1's is dangerous though since there's no guarantee that they only come from external killing due to jobhost restart.

As setting '-1' is not valid, I recommend we should skip the restarts and document this behavior.

**Comment by mhoeger** [comment]
@pragnagopa - definitely agreed that it isn't valid and the java worker shouldn't set the exit code to -1! I'm not sure if you're saying we should bring back the code that ignores a restart on exit code -1, but that's the only part i'd be against (ex: if someone implemented an http worker with logic to exit with -1 for whatever reason).

**Comment by pragnagopa** [comment]
Yes, my recommendation is to bring back the code that ignores restarting on exit code `-1`. Any language/http worker  author should not be setting exit code -1. If they do, it is by design that process will not be restarted.

**Comment by amamounelsayed** [comment]
@pragnagopa Totally agrees that language/http worker should handle the exit -1 case. But before returning back the code should we handle the case that host should not accept new requests in case worker is down or exit by -1.

**Comment by pragnagopa** [comment]
@amamounelsayed - Correct. Issue #https://github.com/Azure/azure-functions-host/issues/5415 is valid and we should fix it. @yojagad  is looking into making some improvements  around this.

**Comment by mhoeger** [comment]
@pragnagopa - i'm still weary of introducing a legitimate (legitimate, meaning it's still bad but there will be no warnings or exceptions) way to get into a hung state because as we saw with the case that prompted this PR, it's extremely impactful. all triggers continue to operate, and all invocations fail for as long as the user does not notice, and once they notice, it requires a manual restart (which is also not necessarily intuitive). i know it's not likely with language workers doing most of the exit code handling, but with the introduction of http workers and the ability for traditional function code to exit in any way they like, it feels like someone will do this (like we already did). i know that -1 is semantically incorrect, but the way a user or potentially language worker author would be punished for making a mistake is pretty painful. What we gain with ignoring exit code -1 is to avoid potentially confusing logs, but i don't think that's worth the trade-off.

**Comment by pragnagopa** [comment]
Unfortunately the root cause is what @amamounelsayed pointed out. If there is no language worker for whatever reason it may be, crashed and not restarted - fix is to shutdown functions host which we already tracking as a separate issue.

**Comment by mhoeger** [comment]
I don't think the root cause is #5415, but I see what you mean! (Added details to 5415, but i think that that issue is tracking the case when grpc server shuts down unexpectedly.)

Should we open a new issue to shut down the host instead of returning each invocation in error for most exceptions thrown in this [try catch block around invocations](https://github.com/Azure/azure-functions-host/blob/aead204f7a7ace3c9bede9f7928249933d18d777/src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs#L233-L250)? I guess all of those should probably be fatal (except perhaps `Function:{invocationContext.FunctionMetadata.Name} is not loaded by the language worker: {rpcWorkerChannel.Id}`)?


Thinking out loud - I think a few things we should be careful about though with more aggressively shutting down in bad states is that it could help mask bugs in our code and there's decreased visibility for customers in what their issue is (if it's perpetually fatal). Ex: when we changed wrong node version and too many worker restarts to be a fatal exception instead of a perpetual invocation exception, we lost the good error message (users now see [a bad gateway error](https://github.com/Azure/azure-functions-host/issues/5163) and more generic messages about the host not being able to start up).

**Pull request merged**